crates/driver/src/domain/competition/solution/settlement.rs

```diff
@@ -268,20 +268,26 @@ impl Settlement {
         auction: &competition::Auction,
         revert_protection: &mempools::RevertProtection,
     ) -> Result<competition::Score, score::Error> {
-        let objective_value = self
-            .boundary
-            .objective_value(eth, auction, self.gas.estimate)?;
+        let quality = self.boundary.quality(eth, auction)?;
 
         let score = match self.boundary.score() {
             competition::SolverScore::Solver(score) => competition::Score(score),
             competition::SolverScore::RiskAdjusted(success_probability) => {
+                let gas_cost = self.gas.estimate * auction.gas_price();
                 // The cost in case of a revert can deviate non-deterministically from the cost
                 // in case of success and it is often significantly smaller. Thus, we go with
                 // the full cost as a safe assumption.
-                let failure_cost =
-                    matches!(revert_protection, mempools::RevertProtection::Disabled)
-                        .then(|| self.gas.estimate * auction.gas_price())
-                        .unwrap_or(zero());
+                let failure_cost = match revert_protection {
+                    mempools::RevertProtection::Enabled => zero(),
+                    mempools::RevertProtection::Disabled => gas_cost,
+                };
+                let objective_value = quality - gas_cost;
+                if objective_value.is_zero() {
```

Comment:
Yes, since negative scores are converted to 0 since U256 is unsigned...

Reply:
If that applies to other parts of the code, then it might be useful to filter out nonpositive  objectives (as in `quality <= gas_costs`). The formulas probably do work as intended in case of objective  0 and result in scores of 0. But there might be edge cases I have not thought about.

Reply:
Do you think I should apply this check in both cases? (solver provided direct score vs solver provided success probability)?

Reply:
With direct score submission, we should probably not check `quality > gas_cost`. At least this is a restriction not mentioned in the CIP and not strictly necessary. For example if our cost estimates are slightly off this could lead to exclusion of good solutions.

Reply:
> With direct score submission, we should probably not check quality > gas_cost.

Ok, this is the current state of the code. For direct score submission, (1) and (2) are checked. 

Another question, if (3) and (4) passed in score calculator, can it ever be true `score > objective_value` ?

Reply:
> Another question, if (3) and (4) passed in score calculator, can it ever be true score > objective_value

Based on the UTs and my understanding, score should be at max equal to objective value, but just want to confirm.

Reply:
If (3) (and (4)) holds then we should always have `score <= objective_value`. I.e. `score > objective_value` is _not_ possible then.

Reply:
We can resolve this as we agreed on mutual understanding;